x/market/handler/endblock_test.go

```diff
@@ -0,0 +1,133 @@
+package handler
+
+import (
+	"errors"
+	"math/rand"
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/ovrclk/akash/testutil"
+	"github.com/ovrclk/akash/x/market/types"
+)
+
+var (
+	// Create and seed the generator.
+	// Typically a non-fixed seed should be used, such as time.Now().UnixNano().
+	// Using a fixed seed will produce the same output on every run.
+	r = rand.New(rand.NewSource(550000))
+)
+
+type winnerTest struct {
+	desc      string
+	bids      []types.Bid
+	expWinner *types.Bid
+	expErr    error
+}
+
+func (w *winnerTest) testFunc(t *testing.T) {
+	winner, err := pickBidWinner(w.bids)
+	if !errors.Is(err, w.expErr) {
+		t.Errorf("returned err: %v does not match %v", err, w.expErr)
+	}
+	if w.expWinner != nil && !winner.Equals(w.expWinner.ID()) {
+		t.Errorf("unexpected winner: %#v\n%q : %v", winner, types.BidIDString(winner.BidID), winner.Price)
+		t.Logf("winner: %+v coin: %s", winner.ID(), winner.Price.String())
+	}
+}
+
+func TestBidWinner(t *testing.T) {
+	seed := make([]byte, 32)
+	_, err := r.Read(seed)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ownerAddr := testutil.DeterministicAccAddress(t, seed)
+
+	bid0 := createBid(t, int64(3))
+	bid0.Owner = ownerAddr
+	bid1 := createBid(t, int64(5))
+	bid1.Owner = ownerAddr
+	bid2 := createBid(t, int64(5))
+	bid2.Owner = ownerAddr
+	bid3 := createBid(t, int64(4))
+	bid3.Owner = ownerAddr
+	bid4 := createBid(t, int64(5))
+	bid4.Owner = ownerAddr
+	bid5 := createBid(t, int64(5))
+	bid5.Owner = ownerAddr
+	bid6 := createBid(t, int64(5))
+	bid6.Owner = ownerAddr
+	bid7 := createBid(t, int64(8))
+	bid7.Owner = ownerAddr
+	bid8 := createBid(t, int64(6))
+	bid8.Owner = ownerAddr
+	bid9 := createBid(t, int64(5))
+	bid9.Owner = ownerAddr
+
+	var winnerTests = []winnerTest{
+		{
+			desc:      "no bids",
+			bids:      []types.Bid{},
+			expWinner: nil,
+			expErr:    errNoBids,
+		},
+		{
+			desc:      "single bid",
+			bids:      []types.Bid{bid9},
+			expWinner: &bid9,
+		},
+		{
+			desc:      "two bids",
+			bids:      []types.Bid{bid9, bid7},
+			expWinner: &bid9,
+		},
+		{
+			desc:      "two matching bids",
+			bids:      []types.Bid{bid9, bid7, bid4},
+			expWinner: &bid4,
+		},
+		{
+			desc:      "all the same bid values",
+			bids:      []types.Bid{bid9, bid9, bid9, bid9, bid9},
+			expWinner: &bid9,
+		},
+		{
+			bids:      []types.Bid{bid9, bid3},
+			expWinner: &bid3,
+		},
+		{
+			bids:      []types.Bid{bid1, bid0, bid9, bid2, bid3, bid4},
```

Comment:
nit: I think it's clearer to have `makeBid()` here in the table instead having `bid0...bidN`

Reply:
If `createBid()` is called `[]types.Bid{createBid(),...}` style, then it's not possible to assert that a Bid in the set of Bidders was an actual winner. Granted the assertion is just checking the `Provider` address.